include/boost/windows_sspi/stream.hpp

```diff
@@ -201,6 +200,53 @@ struct async_read_impl : boost::asio::coroutine {
   std::vector<char> m_input;
 };
 
+// TODO: Move away from this file
+template <typename NextLayer>
+struct async_shutdown_impl : boost::asio::coroutine {
+  async_shutdown_impl(NextLayer& next_layer, detail::sspi_impl& sspi_impl)
+    : m_next_layer(next_layer)
+    , m_sspi_impl(sspi_impl)
+    , m_entry_count(0) {
+  }
+
+  template <typename Self>
+  void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    if (ec) {
+      self.complete(ec);
+      return;
+    }
+
+    ++m_entry_count;
+    auto is_continuation = [this] {
+      return m_entry_count > 1;
+    };
+
+    BOOST_ASIO_CORO_REENTER(*this) {
+      if (m_sspi_impl.shutdown() == detail::sspi_shutdown::state::data_available) {
+          BOOST_ASIO_CORO_YIELD net::async_write(m_next_layer, net::buffer(m_sspi_impl.shutdown.data()), net::transfer_exactly(m_sspi_impl.shutdown.data().size()), std::move(self));
+          self.complete(m_sspi_impl.handshake.last_error());
+          return;
+      }
+
+      if (m_sspi_impl.shutdown() == detail::sspi_shutdown::state::error) {
+        if (!is_continuation()) {
+          BOOST_ASIO_CORO_YIELD {
+            auto e = self.get_executor();
+            net::post(e, [self = std::move(self), ec, length]() mutable { self(ec, length); });
```

Comment:
Note: posting like this "hides" the associated allocator (it's still correct, but possibly not as performant).

Reply:
I wrote this with the kind help of @madmongo1 and I must admit I'm not 100% clear on exactly what is going on here.

I might need to either the use the functionality in `boost::beast` or reimplement it here. This is also duplicated a few places which would be nice to avoid if possible.